contracts/starknet/space/space.cairo

```diff
@@ -0,0 +1,256 @@
+    proposal_threshold.write(_proposal_threshold)
+    voting_strategy.write(_voting_strategy)
+    authenticator.write(_authenticator)
+    next_proposal_nonce.write(1)
+
+    return ()
+end
+
+@external
+func vote{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr : felt}(
+        voter_address : EthAddress, proposal_id : felt, choice : felt) -> ():
+    alloc_locals
+
+    # Verify that the caller is the authenticator contract.
+    authenticator_only()
+
+    let (proposal) = proposal_registry.read(proposal_id)
+    let (current_block) = get_block_number()
+
+    # Make sure proposal is not closed
+    assert_lt(current_block, proposal.end_block)
+
+    # Make sure proposal has started
+    assert_le(proposal.start_block, current_block)
+
+    # Make sure voter has not already voted
+    let (prev_vote) = vote_registry.read(proposal_id, voter_address)
+    if prev_vote.choice != 0:
+        # Voter has already voted!
+        assert 1 = 0
+    end
+
+    let (local params : felt*) = alloc()
+    let (strategy_contract) = voting_strategy.read()
+
+    let (voting_power) = IVotingStrategy.get_voting_power(
+        contract_address=strategy_contract,
+        address=voter_address,
+        at=current_block,
+        params_len=0,
+        params=params)
+
+    if choice == Choice.FOR:
+        let (for) = power_for.read(proposal_id)
+        # TODO: what should we do about the carry
+        let (power, carry) = uint256_add(for, voting_power)
+        power_for.write(proposal_id, power)
+        tempvar range_check_ptr = range_check_ptr
+        tempvar syscall_ptr = syscall_ptr
+        tempvar pedersen_ptr = pedersen_ptr
+        tempvar voting_power = voting_power
+    else:
+        if choice == Choice.AGAINST:
+            let (against) = power_against.read(proposal_id)
+            # TODO: what should we do about the carry
+            let (power, carry) = uint256_add(against, voting_power)
+            power_against.write(proposal_id, power)
+            tempvar range_check_ptr = range_check_ptr
+            tempvar syscall_ptr = syscall_ptr
+            tempvar pedersen_ptr = pedersen_ptr
+            tempvar voting_power = voting_power
+        else:
+            if choice == Choice.ABSTAIN:
+                let (_abstain) = power_abstain.read(proposal_id)
+                # TODO: what should we do about the carry
+                let (power, carry) = uint256_add(_abstain, voting_power)
+                power_abstain.write(proposal_id, power)
+                tempvar range_check_ptr = range_check_ptr
+                tempvar syscall_ptr = syscall_ptr
+                tempvar pedersen_ptr = pedersen_ptr
+                tempvar voting_power = voting_power
+            else:
+                # choice is not a valid choice ?!
+                return ()
+            end
+        end
+    end
+
+    let vote = Vote(choice=choice, voting_power=voting_power)
+    vote_registry.write(proposal_id, voter_address, vote)
```

Comment:
Can you also emit event here?

Reply:
[5997c7b](https://github.com/snapshot-labs/sx-core/pull/47/commits/5997c7b45cce5cac00f09c5bd84fe6fc4b142120) :D